numpy/core/src/common/half.hpp

```diff
@@ -16,48 +20,280 @@ namespace np {
 
 /// Provides a type that implements 16-bit floating point (half-precision).
 /// This type is ensured to be 16-bit size.
+#if !defined(__ARM_FP16_FORMAT_IEEE) || defined(NPY_DOXYGEN_BUILD)
 class Half final {
- public:
-    /// @name Public Constructors
-    /// @{
+  public:
+    /// Returns true if Half has a full native HW support.
+    static constexpr bool kNative = false;
+    /// Returns true if comparison operations handles FP Q/S/NaNs exceptions.
+    static constexpr bool kCompareSignaling = false;
 
     /// Default constructor. initialize nothing.
     Half() = default;
     /// Copy.
-    Half(const Half &r)
+    Half(const Half &r) : data_(r.data_)
+    {}
+    /// Constract from float
+    /// If there are no hardware optimization available, rounding will always
+    /// be set to ties to even.
+    explicit Half(float f)
     {
-        data_.u = r.data_.u;
+    #if defined(NPY_HAVE_FP16)
+        __m128 mf = _mm_load_ss(&f);
+        data_ = static_cast<uint16_t>(_mm_cvtsi128_si32(_mm_cvtps_ph(mf, 0)));
+    #elif defined(NPY_HAVE_VSX3) && defined(NPY_HAVE_VSX_ASM)
+        __vector float vf32 = vec_splats(f);
+        __vector unsigned short vf16;
+        __asm__ __volatile__ ("xvcvsphp %x0,%x1" : "=wa" (vf16) : "wa" (vf32));
+        data_ = vec_extract(vf16, 0);
+    #else
+        data_ = half_private::FromFloat(f);
+    #endif
     }
 
-    /// @}
+    /// Construct from double.
+    /// If there are no hardware optimization available, rounding will always
+    /// be set to ties to even.
```

Comment:
Doesn't this result in a double round-to-even?
Considering a reduced case of floats with 8, 4, and 2 bit of mantissa;
* `01011111` -> `01` (direct rounding from 8 to 2 bits)
* `01011111` -> `0110` -> `10` (rounding via the intermediate 4 bits)

Reply:
Agreed, I reverted the change and reinstated the previous behavior, which relies on explicit conversion instead of using single-precision as an argument when hardware-native conversion is available